### PR TITLE
Allow rustdoc to get compiled with debuginfo

### LINKED ISSUE #44810: rustdoc is not built with debuginfo, even if `debuginfo=true`
While investigating https://github.com/rust-lang/rust/pull/44613, i found that even though i set `debuginfo = true` in my `config.toml`, the symbols that came out would not include any source information for librustdoc. Backtraces for panics didn't have useful method names unless i turned off optimizations. Attempts to debug rustdoc in Visual Studio would just have a blind spot for any source information within librustdoc itself. Profiles would show mangled names for any librustdoc methods.

EDIT: I should note, a `rustdoc-tool-binary-(hash).pdb` was created, and Visual Studio dutifully found it and loaded it, but it still had all the problems noted above.

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Fixes #44810

cc @QuietMisdreavus 
r? @alexcrichton

### CODE REVIEW

File: src/bootstrap/tool.rs [hunk 1]
```diff
@@ -304,6 +304,30 @@ impl Step for Rustdoc {
                                            target,
                                            "build",
                                            "src/tools/rustdoc");
+
+        // Most tools don't get debuginfo, but rustdoc should.
+        cargo.env("RUSTC_DEBUGINFO", self.config.rust_debuginfo.to_string())
+             .env("RUSTC_DEBUGINFO_LINES", self.config.rust_debuginfo_lines.to_string())
```
Comment (kennytm) [inline src/bootstrap/tool.rs]:
```
[00:02:02] error[E0609]: no field `config` on type `tool::Rustdoc`
[00:02:02]    --> /checkout/src/bootstrap/tool.rs:309:43
[00:02:02]     |
[00:02:02] 309 |         cargo.env("RUSTC_DEBUGINFO", self.config.rust_debuginfo.to_string())
[00:02:02]     |                                           ^^^^^^ unknown field
[00:02:02]     |
[00:02:02]     = note: available fields are: `host`
[00:02:02] 
[00:02:02] error[E0609]: no field `config` on type `tool::Rustdoc`
[00:02:02]    --> /checkout/src/bootstrap/tool.rs:310:49
[00:02:02]     |
[00:02:02] 310 |              .env("RUSTC_DEBUGINFO_LINES", self.config.rust_debuginfo_lines.to_string())
[00:02:02]     |                                                 ^^^^^^ unknown field
[00:02:02]     |
[00:02:02]     = note: available fields are: `host`
[00:02:02] 
[00:02:02] error: aborting due to 2 previous errors
[00:02:02] 
[00:02:02] error: Could not compile `bootstrap`.
```

**Comment by QuietMisdreavus** [comment]
librustdoc failed to build, citing the metadata flag:

```
[00:34:22] error[E0523]: found two different crates with name `env_logger` that are not distinguished by differing `-C metadata`. This will result in symbol conflicts between the two.
[00:34:22]   --> /checkout/src/librustdoc/lib.rs:37:1
[00:34:22]    |
[00:34:22] 37 | extern crate rustc_driver;
[00:34:22]    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:34:22] 
[00:34:22] error: Could not compile `rustdoc`.
```

**Comment by QuietMisdreavus** [comment]
I checked out this branch, and can confirm that Visual Studio happily set breakpoints and stepped through code in librustdoc with this change.

File: src/bootstrap/tool.rs [hunk 2]
```diff
@@ -304,6 +304,12 @@ impl Step for Rustdoc {
                                            target,
                                            "build",
                                            "src/tools/rustdoc");
+
+        // Most tools don't get debuginfo, but rustdoc should.
+        cargo.env("RUSTC_DEBUGINFO", builder.config.rust_debuginfo.to_string())
+             .env("RUSTC_DEBUGINFO_LINES", builder.config.rust_debuginfo_lines.to_string())
+             .env("RUSTC_FORCE_UNSTABLE", "1");
```
Comment (alexcrichton) [inline src/bootstrap/tool.rs]: I don't think this env var is necessary because we're not shipping rlibs, just the rustdoc executable

**Comment by alexcrichton** [comment]
r=me with one minor nit

**Comment by Mark-Simulacrum** [comment]
@bors r=alexcrichton p=1

Nominating for beta backport. Tiny patch, and I think without this we may lose nice ICEs from rustdoc. Not overly concerned, though, and I'm not sure how much debug information we ship anyway...

**Comment by alexcrichton** [comment]
I don't think we've had rustdoc debuginfo turned on for quite awhile now, right?

**Comment by alexcrichton** [comment]
(in that I wouldn't personally be inclined to backport)

**Comment by Mark-Simulacrum** [comment]
Yeah, I don't really care about the backport. I'll denominate -- we can always backport if necessary.

**Pull request merged**